fairscale/nn/data_parallel/README_fsdp.md

```diff
@@ -0,0 +1,80 @@
+
+The key insight to unlock full parameter sharding is that we can decompose the [all-reduce](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/collectives.html#allreduce) operation in DDP into separate [all-gather](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/collectives.html#allgather) and [reduce-scatter](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/collectives.html#reducescatter) operations:
+
+<img width="864" alt="Screen Shot 2021-01-12 at 12 35 19 PM" src="https://user-images.githubusercontent.com/231798/108780259-26870a00-7536-11eb-890d-51720f39d098.png">
+
+Then, we can rearrange the reduce-scatter + all-gather so that each DDP worker only needs to store a single shard of parameters and optimizer state. The figure below illustrates standard DDP training (left) and fully sharded training (right):
+
+<img width="1119" alt="Screen Shot 2021-02-24 at 4 39 55 PM" src="https://user-images.githubusercontent.com/231798/109069252-f9199800-76be-11eb-96f8-86767edf1eb9.png">
+
+To maximize memory efficiency we can discard the full weights after each layer's forward pass, saving memory for subsequent layers. This can be implemented by applying the FSDP wrapper to every layer in your network (with `reshard_after_forward=True`). In pseudo-code:
+```
+FSDP forward pass:
+    for layer_i in layers:
+        all-gather full weights for layer_i
+        forward pass for layer_i
+        discard full weights for layer_i
+FSDP backward pass:
+    for layer_i in layers:
+        all-gather full weights for layer_i
+        backward pass for layer_i
+        discard full weights for layer_i
+        reduce-scatter gradients for layer_i
+```
+#### Mixed Precision
+
+When `mixed_precision=True`:
+
+- Sharded parameters are downcast to `fp16` before `forward`, promoted to `fp32` after forward.
+- buffers: batch norm not handled in any special way, buffers are kept in `fp16`. Buffers are not sharded regardless of arguments.
+
+- By default, gradients will be computed and reduced `fp32_reduce_scatter` controls 
+- FIXME: If `torch.amp.autocast` is enabled it will over-ride the output dtypes of some operations
+
+
+#### Using CPU RAM
+
+`move_grads_to_cpu` and `cpu_offload` control which tensors get moved to CPU.
+
+- `cpu_offload` moves weights to CPU when they are not being used. 
+- `move_grads_to_cpu` moves gradients to CPU. The use of this option requires that the optimizer has a copy of the model parameters on CPU.
+
+#### Gradient Clipping
+By default, 
+```python
+sharded_module = FullyShardedDataParallel(my_module)
+torch.nn.utils.clip_grad_norm_(sharded_module.parameters(), max_norm=1.0)
+```
+will use an incorrect norm (the norm over all params in a shard) when clipping gradients.
+To overcome this, you can either call
+`sharded_module.clip_grad_norm(1.0)`
+which does the extra computation required to compute the norm properly, or use `torch.nn.utils.clip_grad_value_`.
+```
+
+
+
+#### Misc
```

Comment:
should this include the new wrap/auto_wrap feature?

Reply:
Yes leave desired `md` in comment else I can take a pass tomorrow.

Reply:
```
#### The `enable_wrap` context

There are two cases where the `enable_wrap` context can be useful:

* When you'd like to apply the same parameters to all child modules that you wrap with FSDP. Calling the `wrap` function within the said context will save you from passing the same set of FSDP parameters explicitly.
* When wrapping large models that does NOT fit within the CPU memory. I.e. you don't first create the full model and then traverse it to wrap it with FSDP at different parts. Instead, you create a wrapped instance of the model incrementally as you build up the model, allowing large modules to be sharded in-place.

example:
        with enable_wrap(**fsdp_params):
            # Wraps layer in FSDP by default if within context
            self.l1 = wrap(torch.nn.Linear(5, 5))
            # Wraps children modules by default based on min_num_params
            self.l2 = auto_wrap(TransformerBlock(), min_num_params=1e8)
```